core/src/main/java/io/undertow/server/handlers/accesslog/DefaultAccessLogReceiver.java

```diff
@@ -309,11 +325,59 @@ public void rotate() {
         }
     }
 
+    @SuppressWarnings("static-access")
     @Override
     public void close() throws IOException {
-        closed = true;
-        if (stateUpdater.compareAndSet(this, 0, 1)) {
-            logWriteExecutor.execute(this);
+        this.closed = true;
+        if (this.stateUpdater.compareAndSet(this, 0, 3)) {
+            flushAndTerminate();
+            return;
+        } else {
+            // state[1,2] - scheduled or running, attempt schedule hijack
+            if (this.stateUpdater.compareAndSet(this, 1, 3)) {
+                //this means this thread raced against scheduled run(). run() will exist ASAP
+                //as 1->2 wont be possible, we are at 3 and this.closed
+                flushAndTerminate();
+                return;
+            }
+            // either failed race to 1->3 or we were in 2. We have to wait here sometime.
```

Comment:
this part needs more thought. Invoking sleep without a check for 2 risks us being in another state. And we need to make this method idempotent, i.e., if it is being closed from 2 different threads, only one of them should run.

Reply:
resolved.